In [1]:
#importing the pandas and numpy libraries
import pandas as pd
import numpy as np

In [20]:
food_data = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding = 'latin-1')
food_data.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

#Question 1: 
##type of error?

In [23]:
my_tuppy=(1,2,5,8)
my_tuppy[2]=6

TypeError: 'tuple' object does not support item assignment

#Question 2:
What will be the output?

In [26]:
### Create a 2nd array
S=[['him','sell'],[90,28,43]]

In [27]:
S[0][1][1]

'e'

The value that was generated as the output based on array indexing is "e"

#Question 3:
Checking the number of rows and columns

In [ ]:
### df.shape

In [ ]:
Question 4:
Which year had the least correlation with ‘Element Code’?

In [36]:
##Selecting the Element Code column and all the years
a=food_data[['Element Code','Y2014','Y2015','Y2016','Y2017','Y2018']]

In [37]:
a.corr(method ='pearson').sort_values('Element Code',ascending=True)

Element Code     Y2014     Y2015     Y2016     Y2017     Y2018
Y2016             0.023444  0.996081  0.995739  1.000000  0.992785  0.992757
Y2015             0.023889  0.994647  1.000000  0.995739  0.988048  0.988208
Y2017             0.024254  0.995230  0.988048  0.992785  1.000000  0.998103
Y2018             0.024279  0.994872  0.988208  0.992757  0.998103  1.000000
Y2014             0.024457  1.000000  0.994647  0.996081  0.995230  0.994872
Element Code      1.000000  0.024457  0.023889  0.023444  0.024254  0.024279

From the results above, Year 2016 had a correlations of 0.023444 against Element code which is the smallest in comaprison to all the other years, Year 2014 had the largest correlation with Element Code

In [ ]:
Question 5:
Which of the following is a python inbuilt module?

In [ ]:
## nath module

Question 6:
Perform a groupby operation on ‘Element’. What is the total number of the sum ofProcessing in 2017?

In [43]:
### here i created series named element2017
element2017=food_data[['Y2017']].groupby(food_data['Element']).sum()

In [44]:
element2017.head(10)

Y2017
Element                                        
Domestic supply quantity             2088198.10
Export Quantity                       182338.80
Fat supply quantity (g/capita/day)     10253.84
Feed                                  223705.68
Food                                 1258888.28
Food supply (kcal/capita/day)         454681.00
Food supply quantity (kg/capita/yr)    48690.04
Import Quantity                       294559.09
Losses                                160614.00
Other uses (non-food)                  91645.97

In [45]:
###filtering only the Processing element  from the data
element2017.loc[['Processing']]

Y2017
Element             
Processing  292836.0

In 2017, there was a sum of of 292836.0 in the processing element in 2017 alone

In [ ]:
Question 7:
Consider the following list of tuples:

In [46]:
## Creating a list of tuples
y=[(2,4),(7,8),(1,5 ,9)]

In [47]:
# asigning an element in the upple to variable x
x = y[1][-1]

In [48]:
x

8

#Question 8:
###Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’. Which of these Areas had the 7th lowest sum in 2017?

In [50]:
area_2017=food_data[['Y2017']].groupby(food_data['Area']).sum()

In [51]:
area_2017.head(10)

Y2017
Area                               
Algeria                   325644.27
Angola                    229159.57
Benin                     124771.22
Botswana                   22101.30
Burkina Faso              101855.07
Cabo Verde                 14650.74
Cameroon                  232030.43
Central African Republic   29937.00
Chad                       71594.68
Comoros                       59.84

In [52]:
sorted_2017_area=area_2017.sort_values('Y2017')  ###sorting the values to get the smallest to largest

In [53]:
sorted_2017_area.iloc[[6],[0]]

Y2017
Area                   
Guinea-Bissau  19102.77

#### position 7 from the least is Guinea Bissau

#Question 9:
###What is the total sum of Wine produced in 2015 and 2018 respectively?

In [57]:
##Grouping Y2015 and Y2018  by Item

item_15_18=food_data[['Y2015','Y2018']].groupby(food_data['Item']).sum()

In [58]:
item_15_18.head(10)

Y2015      Y2018
Item                                         
Alcohol, Non-Food          2180.00    2293.00
Alcoholic Beverages       98783.72   97847.27
Animal Products           11811.73   11578.61
Animal fats              200675.72  269648.27
Apples and products       10559.15    9640.51
Aquatic Animals, Others       9.26       5.92
Aquatic Plants              631.23     452.02
Aquatic Products, Other     638.42     456.65
Bananas                   55046.76   55549.23
Barley and products       29153.44   31253.70

In [59]:
##filtering the wine from the resulting dataframe using the loc method
item_15_18.loc[['Wine']]

Y2015    Y2018
Item                  
Wine  4251.81  4039.32

in Y2015 the value of wine was 4251.81 while in 2018 wine was 4039.32

Question 11:
select column ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’. Which ofthese Areas had the highest sum in 2017?

In [61]:
### Grouping area by the values in 2017
area_2017=food_data[['Y2017']].groupby(food_data['Area']).sum()

In [62]:
area_2017.head(10)

Y2017
Area                               
Algeria                   325644.27
Angola                    229159.57
Benin                     124771.22
Botswana                   22101.30
Burkina Faso              101855.07
Cabo Verde                 14650.74
Cameroon                  232030.43
Central African Republic   29937.00
Chad                       71594.68
Comoros                       59.84

In [63]:
area_2017.sort_values('Y2017',ascending=False).head(10)   ###Used to sort the values from the emerging data

Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92

In [64]:
#### The results shows that Nigeria had the highest values in 2017

In [65]:
#Question 12:
### How do you create a pandas DataFrame using this list, to look like the table below?

In [66]:
###creating 2 lists to be used in buil;ding a dataframe
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]
col = ['Age','Nationality','Overall']

In [67]:
##Creating dataframe from the above lists
pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

In [68]:
#Question 13:
### What is the total number and percentage of missing data in 2014 to 3 decimal

In [70]:
#getting total null values
nulls_2014=food_data['Y2014'].isnull().sum()

In [71]:
nulls_2014

1589

In [73]:
#getting total null + non null values
total=nulls_2014 +food_data['Y2014'].notnull().sum()

In [74]:
total

60943

In [75]:
nulls_2014/total*100

2.6073544131401474

#### The number of null values in Y2014 was 1589,  this wa a 2.6073544131401474% of the total values which is 60943

In [76]:
#Question 14:
### Perform a groupby operation on ‘Element’. What year has the highest sum of StockVariation?

In [77]:
#grouping values in years by the element values
element_by_years=food_data[['Y2014','Y2015','Y2016','Y2017','Y2018']].groupby(food_data['Element']).sum()

In [78]:
element_by_years.head(10)

Y2014       Y2015       Y2016  \
Element                                                                   
Domestic supply quantity             1996716.35  2021493.55  2044842.70   
Export Quantity                       150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)     10225.56    10235.74    10102.77   
Feed                                  216927.89   225050.22   228958.65   
Food                                 1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)         454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)    49650.63    49345.13    48985.28   
Import Quantity                       274144.48   267018.46   286582.78   
Losses                                153223.00   155439.00   157787.00   
Other uses (non-food)                  78718.13    66254.41    69563.68   

                                          Y2017       Y2018  
Element                                                      
Domestic supply quantity             2088198.10  2161192.10  
Export Quantity                       182338.80   181594.80  
Fat supply quantity (g/capita/day)     10253.84    10258.69  
Feed                                  223705.68   233489.68  
Food                                 1258888.28  1303841.28  
Food supply (kcal/capita/day)         454681.00   455261.00  
Food supply quantity (kg/capita/yr)    48690.04    49056.85  
Import Quantity                       294559.09   287997.09  
Losses                                160614.00   163902.00  
Other uses (non-food)                  91645.97    91300.97

In [79]:
#filtering only the Stock Variation
element_by_years.loc[['Stock Variation']]

Y2014     Y2015     Y2016     Y2017     Y2018
Element                                                          
Stock Variation  58749.83  34910.99  33140.12  54316.91  20577.91

In [80]:
#### Under stock variation, Y2014 had the highest with 58749.83

Question 15:
What is the mean and standard deviation across the whole dataset for the year2017 to 2 decimal places?

In [83]:
# The Mean
food_data['Y2017'].mean()

140.917764860268

In [84]:
# The Standard deveiation
food_data['Y2017'].std()

1671.8623590572788

#### The mean is 140.917764860268
#### The standard devaition is  1671.8623590572788

#Question 16:
###What is the total number of unique countries in the dataset?

In [86]:
#This generates a series of Countries
countries=food_data['Area'].unique()

In [87]:
#Getting the shape of the unique array of countries generated 
countries.shape

(49,)

#### The are 49 countries inside the area column of this dataset

Question 17:
Which of the following dataframe methods can be used to access elements acrossrows and columns?

In [ ]:
df.iloc[] and df.loc[] are used to access elements across rows and columns

Question 18:
A pandas Dataframe with dimensions (100,3) has how many features andobservations?

#### 3 features, 100 observations, 
#### features are the column or label names while obsewrvations are the row names

#Question 19
###What is the total Protein supply quantity in Madagascar in 2015?

In [89]:
food_data.head(3)

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   

                         Element             Unit    Y2014    Y2015    Y2016  \
0  Total Population - Both sexes     1000 persons  38924.0  39728.0  40551.0   
1       Domestic supply quantity      1000 tonnes      0.0      0.0      0.0   
2  Food supply (kcal/capita/day)  kcal/capita/day   3377.0   3379.0   3372.0   

     Y2017    Y2018  
0  41389.0  42228.0  
1      0.0      0.0  
2   3341.0   3322.0

In [91]:
madagascar_element=food_data.loc[food_data['Area']=='Madagascar',['Element','Y2015']]

In [92]:
madagascar_element

Element     Y2015
30715           Total Population - Both sexes  24234.00
30716                Domestic supply quantity      0.00
30717           Food supply (kcal/capita/day)   1922.00
30718  Protein supply quantity (g/capita/day)     43.26
30719      Fat supply quantity (g/capita/day)     22.97
...                                       ...       ...
32149                                    Food     11.00
32150     Food supply quantity (kg/capita/yr)      0.44
32151           Food supply (kcal/capita/day)      0.00
32152  Protein supply quantity (g/capita/day)      0.01
32153      Fat supply quantity (g/capita/day)      0.01

[1439 rows x 2 columns]

In [93]:
madagascar_proteins=madagascar_element[madagascar_element['Element']=='Protein supply quantity (g/capita/day)']

In [94]:
madagascar_proteins

Element  Y2015
30718  Protein supply quantity (g/capita/day)  43.26
30721  Protein supply quantity (g/capita/day)  34.04
30724  Protein supply quantity (g/capita/day)   9.23
30740  Protein supply quantity (g/capita/day)  27.26
30756  Protein supply quantity (g/capita/day)   2.14
...                                       ...    ...
32113  Protein supply quantity (g/capita/day)   0.01
32123  Protein supply quantity (g/capita/day)   0.00
32133  Protein supply quantity (g/capita/day)   0.02
32142  Protein supply quantity (g/capita/day)   0.00
32152  Protein supply quantity (g/capita/day)   0.01

[111 rows x 2 columns]

In [95]:
madagascar_proteins[['Y2015']].sum(axis=0)

Y2015    173.05
dtype: float64

#### in 2015, madagascar supply of proteins was   173.05 Protein supply quantity (g/capita/day)	

#Question 20:
###How would you select the elements in bold and italics from the array?

In [96]:
array = ([[94,89, 63],[93,92, 48],[92, 94, 56]])

In [98]:
array

[[94, 89, 63], [93, 92, 48], [92, 94, 56]]